test/test-tty-mouse-tracking.c

```diff
@@ -0,0 +1,1440 @@
+  DWORD written;
+  INPUT_RECORD records[2];
+
+  make_mouse_event_records(tty_out, pos, button_state, ctr_key_state,
+                           event_flags, &records[0]);
+  make_mouse_event_records(tty_out, pos, 0, ctr_key_state,
+                           event_flags, &records[1]);
+  if (click == PRESS || click == BOTH) {
+    WriteConsoleInputW(tty_in->handle, &records[0], 1, &written);
+    ASSERT(written == 1);
+  }
+  if (click == RELEASE || click == BOTH) {
+    WriteConsoleInputW(tty_in->handle, &records[1], 1, &written);
+    ASSERT(written == 1);
+  }
+}
+
+static void write_console_mouse_wheel(uv_tty_t* tty_out, uv_tty_t* tty_in,
+                                      COORD pos, rotate_t rotate) {
+  DWORD written;
+  INPUT_RECORD record;
+
+  /* The value of dwButtonState uses the value confirmed by the debugger. */
+  if (rotate == FORWARD) {
+    /* 0000 0000 0111 1000 0000 0000 0000 0000(7864320) */
+    make_mouse_event_records(tty_out, pos, 7864320, 0,
+                             MOUSE_WHEELED, &record);
+  } else {
+    /* 1111 1111 1000 1000 0000 0000 0000 0000(4287102976) */
+    make_mouse_event_records(tty_out, pos, 4287102976u, 0,
+                             MOUSE_WHEELED, &record);
+
+  }
+  WriteConsoleInputW(tty_in->handle, &record, 1, &written);
+  ASSERT(written == 1);
+}
+
+static int get_fd(const char *name) {
+  int fd;
+  HANDLE handle;
+
+  handle = CreateFileA(name,
+                       GENERIC_READ | GENERIC_WRITE,
+                       FILE_SHARE_READ | FILE_SHARE_WRITE,
+                       NULL,
+                       OPEN_EXISTING,
+                       FILE_ATTRIBUTE_NORMAL,
+                       NULL);
+  ASSERT(handle != INVALID_HANDLE_VALUE);
+  fd = _open_osfhandle((intptr_t) handle, 0);
+  ASSERT(fd >= 0);
+  ASSERT(UV_TTY == uv_guess_handle(fd));
+
+  return fd;
+}
+
+static void initialize_tty(uv_tty_t *tty_in, uv_tty_t *tty_out) {
+  int r, fd;
+
+  /*
+   * FIXME Testing may fail due to key input and mouse input from the actual
+   * console. If possible, it is necessary to block input from the actual
+   * console.
+   */
+  fd = get_fd("conin$");
+  r = uv_tty_init(uv_default_loop(), tty_in, fd, 1);  /* Readable .*/
+  ASSERT(r == 0);
+  ASSERT(uv_is_readable((uv_stream_t*) tty_in));
+  ASSERT(!uv_is_writable((uv_stream_t*) tty_in));
+
+  /* Turn on raw mode. */
+  r = uv_tty_set_mode(tty_in, UV_TTY_MODE_RAW);
+  ASSERT(r == 0);
+
+  r = uv_read_start((uv_stream_t*)tty_in, tty_alloc, tty_read);
+  ASSERT(r == 0);
+
+  fd = get_fd("conout$");
+  r = uv_tty_init(uv_default_loop(), tty_out, fd, 1);  /* Writeable .*/
+  ASSERT(r == 0);
+  ASSERT(!uv_is_readable((uv_stream_t*) tty_out));
+  ASSERT(uv_is_writable((uv_stream_t*) tty_out));
+
+  uv_tty_set_vterm_state(UV_TTY_UNSUPPORTED);
+}
+
+static void initialize_buf(uv_buf_t **buf) {
+  *buf = (uv_buf_t*)malloc(sizeof(uv_buf_t));
+  ASSERT(*buf);
+  (*buf)->len = 0;
+  (*buf)->base = (char*)malloc(BUF_SIZE);
```

Comment:
```suggestion
  (*buf)->base = malloc(BUF_SIZE);
```

Reply:
Casting to suppress `gcc` warnings. Whether to cast the result of `malloc` on a file in the test folder is mixed. If it is a rule of the libuv project not to cast, I will follow it.

Reply:
GCC shouldn't warn about this. AFAIK, `libuv` generally tries to avoid redundant casts such as this:
```
Jameson:libuv jameson$ grep -RI '= uv__malloc' src/ | wc -l
      72
Jameson:libuv jameson$ grep -RI ')uv__malloc' src/ | wc -l
      17
Jameson:libuv jameson$ grep -RI ') uv__malloc' src/ | wc -l
      14
```